src/main/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandler.java

```diff
@@ -177,36 +235,69 @@ public PaymentHandleResponse executePayment(@Nonnull String paypalPlusPaymentId,
-                    } else {
-                        throw new PaypalPlusException(format("Error when approving payment [%s], current state=[%s]",
-                                paypalPlusPaymentId, paypalPayment.getState()));
-                    }
+        PaymentExecution paymentExecution = new PaymentExecution().setPayerId(paypalPlusPayerId);
+
+        AddInterfaceInteraction interactionAction = createAddInterfaceInteractionAction(paymentExecution, REQUEST);
+        return ctpFacade.getPaymentService()
+                // 1. add execute request to the ctp payment as interaction interface
+                .updatePayment(ctpPayment.getId(), Collections.singletonList(interactionAction))
+                .thenCompose(payment -> {
+                    // 2. send execute request to paypal
+                    return paypalPlusFacade.getPaymentService().execute(new Payment().setId(paypalPlusPaymentId),
+                            paymentExecution)
+                            .thenCompose(paypalPayment -> {
+                                AddInterfaceInteraction addInteractionAction = createAddInterfaceInteractionAction(paypalPayment, RESPONSE);
+                                // 3. add paypal response to the ctp payment as interaction interface
+                                return ctpFacade.getPaymentService().updatePayment(payment, Collections.singletonList(addInteractionAction))
+                                        // 4. create charge transaction in the ctp payment
+                                        .thenApply(updatedCtpPayment -> createChargeTransaction(paypalPlusPaymentId, paypalPayment, updatedCtpPayment));
+                            });
                 })
-                .thenApply(payment -> PaymentHandleResponse.ofHttpStatus(HttpStatus.CREATED));
+                .thenApply(ignore -> PaymentHandleResponse.ofHttpStatus(HttpStatus.CREATED));
     }
 
-    private PaymentHandleResponse executeWithExceptionallyHandling(String paymentId,
-                                                                   String paymentIdType,
-                                                                   CompletionStage<PaymentHandleResponse> completionStage) {
-        return completionStage
+    private CompletionStage<io.sphere.sdk.payments.Payment> createChargeTransaction(@Nonnull String paypalPlusPaymentId,
+                                                                                    @Nonnull Payment paypalPayment,
+                                                                                    @Nonnull io.sphere.sdk.payments.Payment ctpPayment) {
+        if (PaypalPlusPaymentStates.APPROVED.equals(paypalPayment.getState())) {
+            return createChargeTransaction(paypalPayment, ctpPayment.getId(), SUCCESS);
+        } else if (PaypalPlusPaymentStates.CREATED.equals(paypalPayment.getState())) {
```

Comment:
what about `PENDING`, like in _50_patch_before_payment_ ?

If the set of the values is reused (e.g., like [here](https://github.com/commercetools/commercetools-paypal-plus-integration/blob/75a992a7730d5782a7d50667181c5ed6e4629abc/src/main/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandler.java#L330-L331)) - we have to create some pre-defined set (like array/list/set of `[CREATED, PENDING]`

Reply:
`PENDING` is in this 50 PR as I forgot to include in this one. So far the set is used only here and I don't see any future usage.